aesara/tensor/blockwise.py

```diff
@@ -0,0 +1,451 @@
+        #         raise ValueError(
+        #             f"Inconsistent size for core dimension {dim}: {size} vs {dim_sizes[dim]}"
+        #         )
+
+
+def _parse_input_dimensions(
+    args: Tuple["TensorVariable", ...], input_core_dims: List[Tuple[str, ...]]
+) -> Tuple[Tuple[Variable, ...], Dict[str, "TensorVariable"]]:
+    """Parse broadcast and core dimensions for vectorize with a signature.
+
+    From `numpy.lib.function_base`.
+
+    Parameters
+    ----------
+    args
+        Tuple of input arguments to examine.
+    input_core_dims
+        List of core dimensions corresponding to each input.
+
+    Returns
+    -------
+    broadcast_shape
+        Common shape to broadcast all non-core dimensions to.
+    dim_sizes
+        Common sizes for named core dimensions.
+    """
+    broadcast_args = []
+    dim_sizes: Dict[str, "TensorVariable"] = {}
+    for arg, core_dims in zip(args, input_core_dims):
+        _update_dim_sizes(dim_sizes, arg, core_dims)
+        ndim = arg.ndim - len(core_dims)
+        arg_shape = shape_tuple(arg)
+        broadcast_args.append(arg_shape[:ndim])
+    bcast_shape = broadcast_shape(*broadcast_args, arrays_are_shapes=True)
+    return bcast_shape, dim_sizes
+
+
+def _calculate_shapes(
+    broadcast_shape: Tuple[Variable, ...],
+    dim_sizes: Dict[str, Variable],
+    list_of_core_dims: List[Tuple[str, ...]],
+) -> List[Tuple[Variable, ...]]:
+    """Helper for calculating broadcast shapes with core dimensions.
+
+    From `numpy.lib.function_base`.
+
+    """
+
+    def get_dim_size(x):
+        res = dim_sizes.get(x)
+
+        if res is None:
+            try:
+                return scalar_constant(int(x))
+            except (ValueError, TypeError):
+                # Return a symbolic placeholder for new dimension references
+                # For example, a signature like `("m", "n") -> ("p",)` means
+                # that there will be no `"p"` label to reference in `dim_sizes`
+                # (i.e. pre-existing dimension labels that already have values
+                # assigned to them).
+                return int64(name=x)
+
+        return res
+
+    return [
+        broadcast_shape + tuple(get_dim_size(dim) for dim in core_dims)
+        for core_dims in list_of_core_dims
+    ]
+
+
+def gufunc_sign_to_str(sign):
+    in_sign = [f"({','.join(_sign)})" for _sign in sign[0]]
+    out_sign = [f"({','.join(_sign)})" for _sign in sign[1]]
+    return f"{','.join(in_sign)}->{','.join(out_sign)}"
+
+
+def infer_shape_to_gufunc_sig(node: Apply, fgraph: Optional["FunctionGraph"] = None):
+    """Derive a gufunc signature from an `Op.infer_shape`.
+
+    Parameters
+    ==========
+    node
+        The `Apply` node with the `Op.infer_shape` we want to use.
+    fgraph
+        A `FunctionGraph` containing `node`.
+
+    """
+    op = node.op
+    in_shapes = tuple(
+        tuple(lscalar(f"i{s}") for s in range(inp.type.ndim)) for inp in node.inputs
```

Comment:
Looks like we need to change the labels to reflect the inputs.  For example:
```suggestion
        tuple(lscalar(f"i{n}{s}") for s in range(inp.type.ndim)) for n, inp in enumerate(node.inputs)
```